lib/imagec/layer_cache.go

```diff
@@ -0,0 +1,151 @@
+// Copyright 2016 VMware, Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package imagec
+
+import (
+	"encoding/json"
+	"fmt"
+	"sync"
+
+	log "github.com/Sirupsen/logrus"
+
+	"github.com/vmware/vic/lib/apiservers/engine/backends/kv"
+	"github.com/vmware/vic/lib/apiservers/portlayer/client"
+	"github.com/vmware/vic/pkg/trace"
+)
+
+// LCache is an in-memory cache to account for existing image layers
+// It is used primarily by imagec when coordinating layer downloads
+// The cache is initially hydrated by way of the image cache at startup
+type LCache struct {
+	m sync.RWMutex
+	// Layers maps from layer id to layer metadata
+	Layers map[string]*ImageWithMeta
+
+	client *client.PortLayer
+}
+
+// LayerNotFoundError is returned when a layer does not exist in the cache
+type LayerNotFoundError struct{}
+
+func (e LayerNotFoundError) Error() string {
+	return "Layer does not exist"
+}
+
+const (
+	layerCacheKey = "layers"
+)
+
+var (
+	layerCache *LCache
+)
+
+func init() {
+	layerCache = &LCache{
+		Layers: make(map[string]*ImageWithMeta),
+	}
+}
+
+// LayerCache returns a reference to the layer cache
+func LayerCache() *LCache {
+	return layerCache
+}
+
+// NewLayerCache will create a new layer cache or rehydrate an
+// existing layer cache from the portlayer k/v store
+func NewLayerCache(client *client.PortLayer) error {
```

Comment:
Calling this NewLayerCache is odd given it's initializing a package wide global - InitializeLayerCache would be more indicative of what's actually occuring.

Most of this code is written as methods on a layerCache object - if it's viable you could have this return an *LCache that is stored into a global by the caller if that's the required usage model. That keeps the globals closer to the code that uses them